docs/coding-guidelines.md

```diff
@@ -47,11 +47,41 @@ When naming configuration structs, use the following guidelines:
 - Use the `Settings` suffix for configuration structs that are set by developers in the code. For example, `component.TelemetrySettings` ends in `Settings` since it is set by developers in the code.
 - Avoid redundant prefixes that are already implied by the package name. For example, use`configgrpc.ClientConfig` instead of `configgrpc.GRPCClientConfig`.
 
-#### Experimental module naming
-
-Experimental modules can be introduced as submodules of stable modules. They MUST have the same name as the stable
-modules prefixed with `x`. For example, `config/confighttp` module can have an experimental module named
-`config/confighttp/xconfighttp` that contains experimental APIs.
+### Module organization
+
+As usual in Go projects, organize your code into packages grouping related functionality. To ensure
+that we can evolve different parts of the API independently, you should also group related packages
+into modules.
+
+We use the following rules for some common situations where we split into separate modules:
+1. Each top-level directory should be a separate module.
+1. Each component referenceable by the OpenTelemetry Collector Builder should be in a separate
+   module. For example, the OTLP receiver is in its own module, different from that of other
+   receivers.
+1. Consider splitting into separate modules if the API may evolve independently in separate groups
+   of packages. For example, the configuration related to HTTP and gRPC evolve independently, so
+   `config/configgrpc` and `config/confighttp` are separate modules.
+1. Testing helpers should be in a separate submodule with the suffix `test`. For example, if you
+   have a module `component`, the helpers should be in `component/componenttest`.
+1. Experimental packages or groups that will later be merged into the another module should be in
+   their own module, named as it will be after integration. For example, if adding profile support
+   to `pdata`, you should add a separate module `pdata/pprofile` for the experimental code.
+1. Experimental code that will be added to an existing stable module can be a submodule with the
+   same name, but prefixed with x. They MUST have the same name as the stable modules prefixed with
+   `x`. For example, `config/confighttp` module can have an experimental module named
+   `config/confighttp/xconfighttp` that contains experimental APIs.
```

Comment:
I find the distinction between points 5 and 6 confusing; if I add experimental profiling code to `pdata`, should I name my experimental module `pdata/pprofile`, or `pdata/xpdata`?

(Also, I assume "prefixed with x" being repeated twice is a mistake)

Reply:
> I find the distinction between points 5 and 6 confusing; if I add experimental profiling code to pdata, should I name my experimental module pdata/pprofile, or pdata/xpdata?

If you are adding a new package/package group you don't add the `x`. If you are adding functionality that you want to eventually add to an existing package, you add the `x`. So for example:
- Profiles are in a separate package, so they don't get the `x`
- Something like #11769 is something we don't want to have in a separate package: the only reason we put it in a separate package is because we can't have experimental and stable API on the same package.

Does that make more sense? If so, how could we rephrase this to make the distinction more clear?

Reply:
That makes sense. I think one way to make this clearer would be:

> 1. Experimental packages or groups ***of packages*** that will later be ***added to a*** module should be in their own module, named as ***they*** will be after integration. For example, if adding ***a `pprofile` package*** to `pdata`, you should add a separate module `pdata/pprofile` for the experimental code.
> 1. Experimental code that will be added to an existing ***package in a*** stable module can be a submodule with the same name, but prefixed with an `x`. For example, `config/confighttp` module can have an experimental module named `config/confighttp/xconfighttp` that contains experimental APIs.

Reply:
Addressed, thanks!